File: server/app/services/composio_service.py
from typing import Dict, Any, Optional, List
from app.config.composio import ComposioConfig

class ComposioService:
    """Service for handling Composio integrations"""
    
    _instance: Optional['ComposioService'] = None
    _client: Optional[Any] = None
    
    def __new__(cls) -> 'ComposioService':
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance
    
    def __init__(self):
        if not hasattr(self, 'initialized'):
            self.initialized = False
            try:
                if ComposioConfig.is_configured():
                    self._initialize_client()
                else:
                    print("Warning: Composio API key not configured. Composio integration will not work.")
            except ImportError:
                print("Warning: Composio package not installed. Composio integration will not work.")
    
    def _initialize_client(self) -> None:
        """Initialize the Composio client"""
        if self.initialized:
            return
            
        try:
            from composio import Composio
            self._client = Composio(
                api_key=ComposioConfig.API_KEY,
                base_url=ComposioConfig.BASE_URL
            )
            self.initialized = True
            print("Composio client initialized successfully")
        except ImportError:
            print("Warning: Composio package not installed. Composio integration will not work.")
            self.initialized = False
        except Exception as e:
            print(f"Failed to initialize Composio client: {e}")
            self.initialized = False
    
    def get_client(self) -> Any:
        """Get the Composio client instance"""
        if not self.initialized or not self._client:
            raise RuntimeError("Composio client not initialized. Check your API key configuration.")
        return self._client
    
    async def initiate_connection(
        self, 
        user_id: str, 
        provider: str, 
        callback_url: Optional[str] = None,
        scope: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """Initiate connection with a provider (OAuth flow)"""
        if not self.initialized or not self._client:
            raise RuntimeError("Composio client not initialized.")
        
        try:
            # Create connection request
            connection_request = self._client.connected_accounts.create(
                user_id=user_id,
                app_name=provider,
                redirect_url=callback_url,
                scope=scope or []
            )
            
            return {
                "connection_id": connection_request.id,
                "redirect_url": connection_request.redirectUrl,
                "status": "initiated"
            }
        except Exception as e:
            raise RuntimeError(f"Failed to initiate connection: {str(e)}")
    
    async def wait_for_connection(
        self, 
        connection_id: str, 
        timeout: int = 60
    ) -> Dict[str, Any]:
        """Wait for a connection to be established"""
        if not self.initialized or not self._client:
            raise RuntimeError("Composio client not initialized.")
        
        try:
            # Wait for connection to be established
            connected_account = self._client.connected_accounts.wait(
                connection_id=connection_id,
                timeout=timeout
            )
            
            return {
                "account_id": connected_account.id,
                "status": connected_account.status,
                "provider": getattr(connected_account, 'appName', ''),
                "created_at": getattr(connected_account, 'createdAt', None)
            }
        except Exception as e:
            raise RuntimeError(f"Failed waiting for connection: {str(e)}")
    
    async def list_connected_accounts(self, user_id: str) -> List[Dict[str, Any]]:
        """List all connected accounts for a user"""
        if not self.initialized or not self._client:
            raise RuntimeError("Composio client not initialized.")
        
        try:
            # Get connected accounts for the user
            response = self._client.connected_accounts.get(user_id=user_id)
            accounts = response.get("items", []) if isinstance(response, dict) else response
            
            result = []
            for account in accounts:
                result.append({
                    "account_id": account.id,
                    "status": account.status,
                    "provider": getattr(account, 'appName', ''),
                    "created_at": getattr(account, 'createdAt', None)
                })
            return result
        except Exception as e:
            raise RuntimeError(f"Failed to list connected accounts: {str(e)}")
    
    async def get_connected_account(self, account_id: str) -> Dict[str, Any]:
        """Get a specific connected account"""
        if not self.initialized or not self._client:
            raise RuntimeError("Composio client not initialized.")
        
        try:
            account = self._client.connected_accounts.get_by_id(account_id=account_id)
            
            return {
                "account_id": account.id,
                "status": account.status,
                "provider": getattr(account, 'appName', ''),
                "created_at": getattr(account, 'createdAt', None),
                "updated_at": getattr(account, 'updatedAt', None)
            }
        except Exception as e:
            raise RuntimeError(f"Failed to get connected account: {str(e)}")
    
    async def enable_connected_account(self, account_id: str) -> bool:
        """Enable a connected account"""
        if not self.initialized or not self._client:
            raise RuntimeError("Composio client not initialized.")
        
        try:
            self._client.connected_accounts.enable(account_id=account_id)
            return True
        except Exception as e:
            raise RuntimeError(f"Failed to enable connected account: {str(e)}")
    
    async def disable_connected_account(self, account_id: str) -> bool:
        """Disable a connected account"""
        if not self.initialized or not self._client:
            raise RuntimeError("Composio client not initialized.")
        
        try:
            self._client.connected_accounts.disable(account_id=account_id)
            return True
        except Exception as e:
            raise RuntimeError(f"Failed to disable connected account: {str(e)}")
    
    async def refresh_connected_account(self, account_id: str) -> bool:
        """Refresh credentials for a connected account"""
        if not self.initialized or not self._client:
            raise RuntimeError("Composio client not initialized.")
        
        try:
            self._client.connected_accounts.refresh(account_id=account_id)
            return True
        except Exception as e:
            raise RuntimeError(f"Failed to refresh connected account: {str(e)}")
    
    async def delete_connected_account(self, account_id: str) -> bool:
        """Delete a connected account"""
        if not self.initialized or not self._client:
            raise RuntimeError("Composio client not initialized.")
        
        try:
            self._client.connected_accounts.delete(account_id=account_id)
            return True
        except Exception as e:
            raise RuntimeError(f"Failed to delete connected account: {str(e)}")
    
    async def execute_tool(
        self, 
        user_id: str, 
        action_name: str, 
        params: Dict[str, Any],
        connected_account_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """Execute a tool/action"""
        if not self.initialized or not self._client:
            raise RuntimeError("Composio client not initialized.")
        
        try:
            # Execute the action
            result = self._client.actions.execute(
                action_name=action_name,
                params=params,
                entity_id=user_id,
                connected_account_id=connected_account_id
            )
            
            return {
                "success": True,
                "data": result
            }
        except Exception as e:
            raise RuntimeError(f"Failed to execute tool {action_name}: {str(e)}")

# Create singleton instance
composio_service = ComposioService()

File: client/src/services/composioService.ts
import { Composio } from "@composio/core";
import { COMPOSIO_CONFIG, type ProviderName } from "../config/composio";

// Re-export ProviderName for convenience
export type { ProviderName } from "../config/composio";

// Define types for Composio responses
interface ConnectedAccount {
  id: string;
  status: string;
  provider: string;
  createdAt: string;
}

interface Tool {
  slug: string;
  name: string;
  description?: string;
}

interface ToolExecutionResult {
  success: boolean;
  data?: unknown;
  error?: string;
}

interface ConnectionResponse {
  connectionId: string;
  redirectUrl: string;
}

/**
 * ComposioService - Service for handling Composio integrations
 */
class ComposioService {
  private composio: Composio | null = null;
  private initialized = false;

  /**
   * Initialize the Composio client
   */
  async initialize(): Promise<void> {
    if (this.initialized) {
      return;
    }

    if (!COMPOSIO_CONFIG.API_KEY) {
      throw new Error(
        "Composio API key is missing. Please set VITE_COMPOSIO_API_KEY in your environment variables."
      );
    }

    try {
      this.composio = new Composio({
        apiKey: COMPOSIO_CONFIG.API_KEY,
        baseURL: COMPOSIO_CONFIG.BASE_URL,
      });

      this.initialized = true;
      console.log("Composio client initialized successfully");
    } catch (error) {
      console.error("Failed to initialize Composio client:", error);
      throw error;
    }
  }

  /**
   * Get the Composio client instance
   */
  getClient(): Composio {
    if (!this.initialized || !this.composio) {
      throw new Error(
        "Composio client not initialized. Call initialize() first."
      );
    }
    return this.composio;
  }

  /**
   * Initiate connection with a provider (OAuth flow)
   * @param userId - User identifier
   * @param provider - Provider name (e.g., 'github', 'slack')
   * @param options - Connection options
   */
  async initiateConnection(
    userId: string,
    provider: ProviderName,
    options?: {
      callbackUrl?: string;
      scope?: string[];
    }
  ): Promise<ConnectionResponse> {
    await this.initialize();

    if (!this.composio) {
      throw new Error("Composio client not available");
    }

    try {
      const connectionRequest = await this.composio.connectedAccounts.initiate(
        userId,
        provider,
        {
          callbackUrl: options?.callbackUrl,
          // Remove config with scope for now as it's not supported
          // config: {
          //   scope: options?.scope,
          // },
        }
      );

      return {
        connectionId: connectionRequest.id,
        redirectUrl: connectionRequest.redirectUrl || "",
      };
    } catch (error) {
      console.error(`Failed to initiate connection with ${provider}:`, error);
      throw error;
    }
  }

  /**
   * Wait for a connection to be established
   * @param connectionId - Connection request ID
   * @param timeout - Timeout in milliseconds (default: 60 seconds)
   */
  async waitForConnection(
    connectionId: string,
    timeout: number = COMPOSIO_CONFIG.CONNECTION_TIMEOUT
  ): Promise<ConnectedAccount> {
    await this.initialize();

    if (!this.composio) {
      throw new Error("Composio client not available");
    }

    try {
      const connectedAccount =
        await this.composio.connectedAccounts.waitForConnection(
          connectionId,
          timeout
        );

      return {
        ...connectedAccount,
        provider: connectedAccount.toolkit?.slug || "unknown",
      } as ConnectedAccount;
    } catch (error) {
      console.error("Failed waiting for connection:", error);
      throw error;
    }
  }

  /**
   * List all connected accounts for a user
   * @param userId - User identifier
   */
  async listConnectedAccounts(userId: string): Promise<ConnectedAccount[]> {
    await this.initialize();

    if (!this.composio) {
      throw new Error("Composio client not available");
    }

    try {
      const accounts = await this.composio.connectedAccounts.list({
        userIds: [userId],
      });

      return (accounts.items || []).map((account) => ({
        ...account,
        provider: account.toolkit?.slug || "unknown",
      })) as ConnectedAccount[];
    } catch (error) {
      console.error("Failed to list connected accounts:", error);
      throw error;
    }
  }

  /**
   * Get a specific connected account
   * @param accountId - Connected account ID
   */
  async getConnectedAccount(accountId: string): Promise<ConnectedAccount> {
    await this.initialize();

    if (!this.composio) {
      throw new Error("Composio client not available");
    }

    try {
      const account = await this.composio.connectedAccounts.get(accountId);
      return {
        ...account,
        provider: account.toolkit?.slug || "unknown",
      } as ConnectedAccount;
    } catch (error) {
      console.error("Failed to get connected account:", error);
      throw error;
    }
  }

  /**
   * Enable a connected account
   * @param accountId - Connected account ID
   */
  async enableConnectedAccount(accountId: string): Promise<void> {
    await this.initialize();

    if (!this.composio) {
      throw new Error("Composio client not available");
    }

    try {
      await this.composio.connectedAccounts.enable(accountId);
    } catch (error) {
      console.error("Failed to enable connected account:", error);
      throw error;
    }
  }

  /**
   * Disable a connected account
   * @param accountId - Connected account ID
   */
  async disableConnectedAccount(accountId: string): Promise<void> {
    await this.initialize();

    if (!this.composio) {
      throw new Error("Composio client not available");
    }

    try {
      await this.composio.connectedAccounts.disable(accountId);
    } catch (error) {
      console.error("Failed to disable connected account:", error);
      throw error;
    }
  }

  /**
   * Refresh credentials for a connected account
   * @param accountId - Connected account ID
   */
  async refreshConnectedAccount(accountId: string): Promise<void> {
    await this.initialize();

    if (!this.composio) {
      throw new Error("Composio client not available");
    }

    try {
      await this.composio.connectedAccounts.refresh(accountId);
    } catch (error) {
      console.error("Failed to refresh connected account:", error);
      throw error;
    }
  }

  /**
   * Delete a connected account
   * @param accountId - Connected account ID
   */
  async deleteConnectedAccount(accountId: string): Promise<void> {
    await this.initialize();

    if (!this.composio) {
      throw new Error("Composio client not available");
    }

    try {
      await this.composio.connectedAccounts.delete(accountId);
    } catch (error) {
      console.error("Failed to delete connected account:", error);
      throw error;
    }
  }

  /**
   * Get available tools for a provider
   * @param provider - Provider name
   */
  async getProviderTools(provider: ProviderName): Promise<Tool[]> {
    await this.initialize();

    if (!this.composio) {
      throw new Error("Composio client not available");
    }

    try {
      // This is a simplified approach - in practice, you might want to filter
      // tools based on the provider or use specific toolkits
      // TODO: Fix this based on actual Composio SDK API
      // const tools = await this.composio.actions.list({
      //   appName: provider,
      // });
      const tools = { items: [] };

      return (tools.items || []) as Tool[];
    } catch (error) {
      console.error(`Failed to get tools for ${provider}:`, error);
      throw error;
    }
  }

  /**
   * Execute a tool/action
   * @param userId - User identifier
   * @param actionName - Name of the action to execute
   * @param params - Parameters for the action
   * @param connectedAccountId - Optional connected account ID
   */
  async executeTool(
    _userId: string,
    _actionName: string,
    _params: Record<string, unknown>,
    _connectedAccountId?: string
  ): Promise<ToolExecutionResult> {
    await this.initialize();

    if (!this.composio) {
      throw new Error("Composio client not available");
    }

    try {
      // TODO: Fix this based on actual Composio SDK API
      // const result = await this.composio.actions.execute(
      //   userId,
      //   actionName,
      //   params,
      //   {
      //     connectedAccountId,
      //   }
      // );
      const result = { success: true, data: null };

      return {
        success: true,
        data: result,
      };
    } catch (error) {
      console.error(`Failed to execute tool ${_actionName}:`, error);
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

// Export singleton instance
export const composioService = new ComposioService();
